Retry Mechanism for Package, DNS and Command Tests

j1shnu opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: ./CONTRIBUTING.md
-->

<!-- _Please make sure to review and check all of these items:_ -->


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make test-all` (UNIX) passes. CI will also test this
- [x] unit and/or integration tests are included (if applicable)
- [x] documentation is changed or added (if applicable)

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Description of change
<!--
Please provide a description of the change here. Be sure to use issue references when applicable:
https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords
-->
Added retry mechanism for package/DNS/command testing. Details are mentioned in issue [#1034](https://github.com/goss-org/goss/issues/1034). The goss file syntax will be the following:

````yaml
package:
  httpd:
    retry_count: 1      # Enables retry mechanism
    retry_delay: 10000  # Delay in milliseconds before retrying the check; duration strings like 10s also work
    installed: true
    name: httpd
    versions:
      - 2.2.15
    skip: false

dns:
  CNAME:c.dnstest.io:
    resolvable: true
    server: 208.67.222.222
    addrs:
      - "a.dnstest.io."
    retry_count: 2      # Enables retry mechanism
    retry_delay: 500ms  # Delay before retrying the check; bare numeric values are milliseconds

command:
  'go version':
    exit-status: 0
    exec: "go version"
    stdout:
      - go version go1.6 linux/amd64
    stderr: []
    timeout: 10000
    skip: false
    retry_count: 1      # Enables retry mechanism
    retry_delay: 5s     # Delay before each retry attempt; bare numeric values are milliseconds
````


<!-- readthedocs-preview goss start -->
----
📚 Documentation preview 📚: https://goss--1037.org.readthedocs.build/en/1037/

<!-- readthedocs-preview goss end -->

j1shnu committed:
Added retry logic in code

j1shnu committed:
Retry logic update

j1shnu committed:
Updated doc

j1shnu committed:
Added retry mechanism for DNS and updated related docs

j1shnu committed:
Removed unwanted printing

ns-mjames commented:
@aelsabbahy Could you help merge this pr? This is imperative for our testing env

j1shnu committed:
Implement retry mechanism in Command validation with configurable retry count and delay

j1shnu committed:
Updated doc command retry mechanism

j1shnu committed:
Merge pull request #1 from j1shnu/issue/1034

Added retry for command executions

j1shnu committed:
code formatting

j1shnu committed:
Merge branch 'issue/1034'

petemounce commented:
I've hit the workflow-approve button for the github actions, but Travis is still a required check and I cannot address its not running.

Please would you add some test-coverage to exercise this code-path? Retries as a feature can become quite fiddly because the feature tends to grow as people decide they want also "exponential back-off" "... with jitter" "... conditional on x" and so on - is there a best-of-breed golang library that might be used instead of a bespoke implementation that this contribution begins? My thought is longer-tail support would be reduced if so, and the testing burden I'm asking for would probably be reduced were a library used instead of something bespoke.

Also, what do you think about making the delay interval units be milliseconds to support finer-grained intervals than seconds - or is this not a concern as the input will accept floats like `0.5`? My thought is that being limited to 1s as the finest grained interval would be limiting in cases where goss tests are used as a health-check expecting to evaluate several times a minute.

j1shnu committed:
Refactor retry mechanism to support ms and added testing for the retry mechanism

j1shnu commented:
Thanks @petemounce, these are helpful suggestions.

I’ve added focused test coverage for the retry path. The new tests cover `retry_delay` parsing/serialization and retry behavior for `package`, `dns`, and `command`, including sub second retry intervals.

On the implementation approach: I kept it in-tree for now rather than adding a retry dependency. The current behavior is still fixed-attempt/fixed-delay only, so I consolidated the retry logic into a shared helper to reduce duplication. If the project would strongly prefer a library here, I’m happy to revisit that.

On delay units: I updated `retry_delay` so bare numeric values are interpreted as milliseconds, and it also accepts Go style duration strings such as `500ms`, `2s`, and `1m`. That gives finer-grained retry intervals for health check style usage. Bare floats are also treated as milliseconds, although explicit duration strings are probably the clearest form going forward.

I don’t think I can address the Travis required check issue from the fork side; it looks like repository side required check configuration rather than a failing code path in the PR.

kgaughan commented on resource/retry_test.go:
```diff
@@ -0,0 +1,309 @@
+	dns := &DNS{
+		id:         "localhost",
+		Resolvable: true,
+		Addrs:      []interface{}{"127.0.0.1"},
```
```suggestion
		Addrs:      []any{"127.0.0.1"},
```

kgaughan commented on resource/retry_test.go:
```diff
@@ -0,0 +1,309 @@
+		id:         "echo ok",
+		Exec:       "echo ok",
+		ExitStatus: 0,
+		Stdout:     []interface{}{"ok"},
```
```suggestion
		Stdout:     []any{"ok"},
```

kgaughan commented on resource/retry_test.go:
```diff
@@ -0,0 +1,309 @@
+	pkg := &Package{
+		id:         "nginx",
+		Installed:  true,
+		Versions:   []interface{}{"1.0.0"},
```
```suggestion
		Versions:   []any{"1.0.0"},
```

kgaughan reviewed (commented):
This is just a few minor nits I spotted on a quick scan of the code. I'll post something a bit more substantial tomorrow.

kgaughan commented:
The code looks fine. I applied the fixes for the nits myself. I'll kick off the tests now to make sure everything's still good.
